### ISSUE TITLE
TCP input plugin does not work properly on Windows Server 2019

### ISSUE THREAD

**Issue by NakagawaMakoto** [fluent/fluent-bit#6344]
## Bug Report

**Describe the bug**
It looks to me that the tcp input plugin does not work properly on Windows. Is there anyone having the same problem?

**To Reproduce**
I send test mesages as follows, but the messages are not handed over to stdout plugin.
> powershell .\fluentbittest.ps1 -report -loop 3

```
PS C:\windows\system32> E:\downloads\fluent-bit-2.0.3-win64\fluent-bit-2.0.3-win64\bin\fluent-bit -i tcp -p listen=127.0.0.1 -o stdout -vvv
Fluent Bit v2.0.3
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/11/05 19:02:12] [ info] Configuration:
[2022/11/05 19:02:12] [ info]  flush time     | 1.000000 seconds
[2022/11/05 19:02:12] [ info]  grace          | 5 seconds
[2022/11/05 19:02:12] [ info]  daemon         | 0
[2022/11/05 19:02:12] [ info] ___________
[2022/11/05 19:02:12] [ info]  inputs:
[2022/11/05 19:02:12] [ info]      tcp
[2022/11/05 19:02:12] [ info] ___________
[2022/11/05 19:02:12] [ info]  filters:
[2022/11/05 19:02:12] [ info] ___________
[2022/11/05 19:02:12] [ info]  outputs:
[2022/11/05 19:02:12] [ info]      stdout.0
[2022/11/05 19:02:12] [ info] ___________
[2022/11/05 19:02:12] [ info]  collectors:
[2022/11/05 19:02:12] [ info] [fluent bit] version=2.0.3, commit=86250fa62d, pid=6256
[2022/11/05 19:02:12] [debug] [engine] coroutine stack size: 98302 bytes (96.0K)
[2022/11/05 19:02:12] [ info] [output:stdout:stdout.0] worker #0 started
[2022/11/05 19:02:12] [ info] [storage] ver=1.3.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2022/11/05 19:02:12] [ info] [cmetrics] version=0.5.4
[2022/11/05 19:02:12] [ info] [ctraces ] version=0.2.5
[2022/11/05 19:02:12] [ info] [input:tcp:tcp.0] initializing
[2022/11/05 19:02:12] [ info] [input:tcp:tcp.0] storage_strategy='memory' (memory only)
[2022/11/05 19:02:12] [debug] [tcp:tcp.0] created event channels: read=1064 write=1068
[2022/11/05 19:02:12] [debug] [downstream] listening on 127.0.0.1:5170
[2022/11/05 19:02:12] [debug] [stdout:stdout.0] created event channels: read=1076 write=1080
[2022/11/05 19:02:12] [ info] [sp] stream processor started
[2022/11/05 19:02:16] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp.c:50] new TCP connection arrived FD=1168
[2022/11/05 19:02:19] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:220] read()=57 pre_len=0 now_len=57
[2022/11/05 19:02:19] [debug] [input chunk] update output instances with new chunk size diff=78
[2022/11/05 19:02:19] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:214] fd=1168 closed connection
```

"E:\downloads\fluent-bit-2.0.3-win64\fluent-bit-2.0.3-win64\bin\fluent-bit -i dummy -o stdout" works just fine.

- Steps to reproduce the problem:
My testing script, fluentbittest.ps1, is as simple as follows;
```
param(
  [hashtable]$message=@{"message"="dummy"},
  [int]$interval=1,
  [int]$loop=10,
  [switch]$report=$false,
  [int]$port=5170
)

$TcpClient=New-Object System.Net.Sockets.TcpClient([IPAddress]::Loopback, $port)
$GetStream = $TcpClient.GetStream()
$StreamWriter = New-Object System.IO.StreamWriter $GetStream

for ($i=0; $i -lt $loop; $i++){
  if ($report) {convertto-json $message -depth 2}
  $StreamWriter.Write($(convertto-json $message -depth 2 -compress))
  start-sleep $interval
}
$StreamWriter.Dispose()
$GetStream.Dispose()
$TcpClient.Dispose()
```

**Expected behavior**
With fluentbit 1.9.9 it works all right, and test messages are handed over to the stdout plugin.

```
PS C:\windows\system32> E:\downloads\fluent-bit-1.9.9-win64\fluent-bit-1.9.9-win64\bin\fluent-bit -i tcp -p listen=127.0.0.1 -o stdout -vvv
Fluent Bit v1.9.9
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/11/05 19:03:46] [ info] Configuration:
[2022/11/05 19:03:46] [ info]  flush time     | 1.000000 seconds
[2022/11/05 19:03:46] [ info]  grace          | 5 seconds
[2022/11/05 19:03:46] [ info]  daemon         | 0
[2022/11/05 19:03:46] [ info] ___________
[2022/11/05 19:03:46] [ info]  inputs:
[2022/11/05 19:03:46] [ info]      tcp
[2022/11/05 19:03:46] [ info] ___________
[2022/11/05 19:03:46] [ info]  filters:
[2022/11/05 19:03:46] [ info] ___________
[2022/11/05 19:03:46] [ info]  outputs:
[2022/11/05 19:03:46] [ info]      stdout.0
[2022/11/05 19:03:46] [ info] ___________
[2022/11/05 19:03:46] [ info]  collectors:
[2022/11/05 19:03:46] [ info] [fluent bit] version=1.9.9, commit=5c03b2e555, pid=6836
[2022/11/05 19:03:46] [debug] [engine] coroutine stack size: 98302 bytes (96.0K)
[2022/11/05 19:03:46] [ info] [output:stdout:stdout.0] worker #0 started
[2022/11/05 19:03:46] [ info] [storage] version=1.3.0, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/11/05 19:03:46] [ info] [cmetrics] version=0.3.7
[2022/11/05 19:03:46] [debug] [tcp:tcp.0] created event channels: read=1064 write=1068
[2022/11/05 19:03:46] [ info] [input:tcp:tcp.0] listening on 127.0.0.1:5170
[2022/11/05 19:03:46] [debug] [stdout:stdout.0] created event channels: read=1076 write=1080
[2022/11/05 19:03:46] [debug] [router] default match rule tcp.0:stdout.0
[2022/11/05 19:03:46] [ info] [sp] stream processor started
[2022/11/05 19:03:50] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp.c:47] new TCP connection arrived FD=1152
[2022/11/05 19:03:53] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:210] read()=57 pre_len=0 now_len=57
[2022/11/05 19:03:53] [debug] [input chunk] update output instances with new chunk size diff=78
[2022/11/05 19:03:53] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:204] fd=1152 closed connection
[2022/11/05 19:03:53] [debug] [task] created task=0000013535E8EB30 id=0 OK
[2022/11/05 19:03:53] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] tcp.0: [1667642633.666685400, {"message"=>"dummy"}]
[1] tcp.0: [1667642633.666690100, {"message"=>"dummy"}]
[2] tcp.0: [1667642633.666690700, {"message"=>"dummy"}]
[2022/11/05 19:03:53] [debug] [out flush] cb_destroy coro_id=0
[2022/11/05 19:03:53] [debug] [task] destroy task=0000013535E8EB30 (task_id=0)
```

**Screenshots**

**Your Environment**
I have tested on Windows Server 2019.

**Additional context**

**Comment by leonardo-albertovich** [comment]
I just reproduced your test in Windows 2019 and the only thing I noticed was that since you are not flushing the stream writer the `start-sleep $interval` causes fluent-bit to receive the entire lot at the end when the connection is closed.

I verified that fluent-bit processed and routed the 10 messages and from what I can see in your 1.9 run the timestamp suggests that those are received and processed at the same time.

I modified your powershell code to make the loop look like this and then fluent-bit got the messages on time as expected : 

```
for ($i=0; $i -lt $loop; $i++){
  if ($report) {convertto-json $message -depth 2}
  $StreamWriter.Write($(convertto-json $message -depth 2 -compress))
  $StreamWriter.Flush()
  start-sleep $interval
}
```

**Comment by NakagawaMakoto** [comment]
Thank you @leonardo-albertovich, I see that behaivour changes flushing properly. But still 2.0.3 does not look working properly.  I have tested with the following on the sending side;

> powershell .\fluentbittest.ps1 -report -loop 2; powershell .\fluentbittest.ps1 -report -loop 2

Here is the log from 1.9.9.
```
[2022/11/07 20:00:59] [ info] [output:stdout:stdout.0] worker #0 started
[2022/11/07 20:00:59] [ info] [storage] version=1.3.0, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/11/07 20:00:59] [ info] [cmetrics] version=0.3.7
[2022/11/07 20:00:59] [debug] [tcp:tcp.0] created event channels: read=1064 write=1068
[2022/11/07 20:00:59] [ info] [input:tcp:tcp.0] listening on 127.0.0.1:5170
[2022/11/07 20:00:59] [debug] [stdout:stdout.0] created event channels: read=1076 write=1080
[2022/11/07 20:00:59] [debug] [router] default match rule tcp.0:stdout.0
[2022/11/07 20:00:59] [ info] [sp] stream processor started
[2022/11/07 20:01:22] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp.c:47] new TCP connection arrived FD=1152
[2022/11/07 20:01:22] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:210] read()=19 pre_len=0 now_len=19
[2022/11/07 20:01:22] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/11/07 20:01:22] [debug] [task] created task=0000025881870220 id=0 OK
[2022/11/07 20:01:22] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] tcp.0: [1667818882.661808700, {"message"=>"dummy"}]
[2022/11/07 20:01:22] [debug] [out flush] cb_destroy coro_id=0
[2022/11/07 20:01:22] [debug] [task] destroy task=0000025881870220 (task_id=0)
[2022/11/07 20:01:23] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:210] read()=19 pre_len=0 now_len=19
[2022/11/07 20:01:23] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/11/07 20:01:23] [debug] [task] created task=0000025881870220 id=0 OK
[2022/11/07 20:01:23] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] tcp.0: [1667818883.669598700, {"message"=>"dummy"}]
[2022/11/07 20:01:23] [debug] [out flush] cb_destroy coro_id=1
[2022/11/07 20:01:23] [debug] [task] destroy task=0000025881870220 (task_id=0)
[2022/11/07 20:01:24] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:204] fd=1152 closed connection
[2022/11/07 20:01:25] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp.c:47] new TCP connection arrived FD=1152
[2022/11/07 20:01:25] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:210] read()=19 pre_len=0 now_len=19
[2022/11/07 20:01:25] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/11/07 20:01:25] [debug] [task] created task=000002588186F7A0 id=0 OK
[2022/11/07 20:01:25] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] tcp.0: [1667818885.219639600, {"message"=>"dummy"}]
[2022/11/07 20:01:25] [debug] [out flush] cb_destroy coro_id=2
[2022/11/07 20:01:25] [debug] [task] destroy task=000002588186F7A0 (task_id=0)
[2022/11/07 20:01:26] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:210] read()=19 pre_len=0 now_len=19
[2022/11/07 20:01:26] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/11/07 20:01:26] [debug] [task] created task=000002588186EEA0 id=0 OK
[2022/11/07 20:01:26] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] tcp.0: [1667818886.222905200, {"message"=>"dummy"}]
[2022/11/07 20:01:26] [debug] [out flush] cb_destroy coro_id=3
[2022/11/07 20:01:26] [debug] [task] destroy task=000002588186EEA0 (task_id=0)
[2022/11/07 20:01:27] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:204] fd=1152 closed connection
```
Here is the log from 2.0.3. You can see that only the first two messages are handed over to stdout. With 1.9.9, all four messages are properly handed over to stdout.
```
[2022/11/07 20:04:42] [ info] [sp] stream processor started
[2022/11/07 20:04:46] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp.c:50] new TCP connection arrived FD=1168
[2022/11/07 20:04:46] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:220] read()=19 pre_len=0 now_len=19
[2022/11/07 20:04:46] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/11/07 20:04:47] [debug] [task] created task=000001375C0DD070 id=0 OK
[2022/11/07 20:04:47] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] tcp.0: [1667819086.172167300, {"message"=>"dummy"}]
[2022/11/07 20:04:47] [debug] [out flush] cb_destroy coro_id=0
[2022/11/07 20:04:47] [debug] [task] destroy task=000001375C0DD070 (task_id=0)
[2022/11/07 20:04:47] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:220] read()=19 pre_len=0 now_len=19
[2022/11/07 20:04:47] [debug] [input chunk] update output instances with new chunk size diff=26
[2022/11/07 20:04:48] [debug] [task] created task=000001375C0DE1F0 id=0 OK
[2022/11/07 20:04:48] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] tcp.0: [1667819087.185552700, {"message"=>"dummy"}]
[2022/11/07 20:04:48] [debug] [out flush] cb_destroy coro_id=1
[2022/11/07 20:04:48] [debug] [task] destroy task=000001375C0DE1F0 (task_id=0)
[2022/11/07 20:04:48] [trace] [input:tcp:tcp.0 at C:\projects\fluent-bit-2e87g\plugins\in_tcp\tcp_conn.c:214] fd=1168 closed connection
```

**Comment by leonardo-albertovich** [comment]
I'm still having trouble reproducing your results, even if I remove the flush line and run your script twice with the exact same line you used I'm getting the four records.

**Comment by NakagawaMakoto** [comment]
Hi @leonardo-albertovich,

I have verified that the recently released fluent-bit-2.0.4-win64.zip works as expected. I do not know what has changed since 2.0.3. I will close this issue and hope it will be working in the future.
